feat: Add DataTable component

mikemurray opened the pull request:
Resolves #83
Impact: **minor**  
Type: **feature**

## Component

DataTable component for displaying tabular data.

## Screenshots
![image](https://user-images.githubusercontent.com/42217/65453459-14acaf00-ddf8-11e9-8f28-1fe4bac98aad.png)


## Breaking changes

none.

## Testing
1. See the docs for DataTable https://deploy-preview-103--heuristic-perlman-0ef024.netlify.com/#/Components/Content/DataTable
1. Play around and ensure things like pagination, and filtering work as expected

mikemurray commented:
Things not available in this table
- Loading state
- empty state
- column sorting (by clicking headers)
- mobile support
- custom filter controls
- row interactions like clicking the entire row to link to something. Custom cell rendered can be used to make things inside cells clickable, but the entire row is not interactive)
- Checkbox icon design is not in line with designs. (this should be a global theme prop override to change the checked icon component)
- Dropdown page selector has a vertical divider. (Best solved in its own PR as hiding the `|` line should be a function of the `Select` component itself)

machikoyasuda commented:
Tested and approved:
- Pagination: Paging forward, backward
- Pagination: Changing the number of rows
- Filter: Typing in `jcb`, a name or a specific email address
- Overall typography
- Overall hover color behavior

Requesting change:
- Add `aria-label` to checkboxes - https://material-ui.com/components/checkboxes/#accessibility

machikoyasuda commented on package/src/components/DataTable/DataTable.js:
```diff
@@ -0,0 +1,350 @@
+const useStyles = makeStyles((theme) => ({
+  pagination: {
+    color: theme.palette.colors.coolGrey500,
+    letterSpacing: 0.28,
```
You can use `captionLetterSpacing` theme variable here.

mikemurray replied:
I think this could be the `<Typography variant=body2 />` as its the best match (if not an exact match for the letter spacing, color, and font size)

machikoyasuda replied:
or that, too!

machikoyasuda commented on package/src/components/DataTable/DataTable.js:
```diff
@@ -0,0 +1,350 @@
+  },
+  tableHead: {
+    fontWeight: theme.typography.fontWeightSemiBold,
+    letterSpacing: 0.5,
```
`h5LetterSpacing` variable

mikemurray replied:
@cassytaylor should the table head just be an H5 (in the display, but as a literal H5 element) and If so, should all H5's be semi-bold?

(trying to reduce one-off font variants like these)

machikoyasuda commented on package/src/components/DataTable/DataTable.js:
```diff
@@ -0,0 +1,350 @@
+    color: theme.palette.colors.coolGrey500
+  },
+  tableCell: {
+    letterSpacing: 0.28,
```
`captionLetterSpacing`

mikemurray replied:
This seems to better match the font-style of body2 as the letter spacing and color are the same as that variant.

machikoyasuda commented on package/src/components/DataTable/DataTable.md:
````diff
@@ -0,0 +1,284 @@
+TableExample()
+```
+
+##### Simple table with data fetching
````
Remove `Simple`.

machikoyasuda commented on package/src/components/DataTable/DataTable.md:
```diff
@@ -0,0 +1,284 @@
+
+##### Simple table with data fetching
+
+Simple table with no filtering or selectable rows. Data fetching is simulated with a 200ms delay.
```
Remove `simple`

machikoyasuda commented on package/src/components/DataTable/DataTable.md:
```diff
@@ -0,0 +1,284 @@
+
+#### Types
+
+##### Standard selectable table data fetching and filtering
```
This title is a little bit of a mouthful. Maybe we can just call this the `Standard table`. The paragraph below can explain that the `Standard table` has selectable rows, data-fetching and filtering.

mikemurray commented:
@machikoyasuda Ready for a re-review.

For the aria labels, react-table automatically sets the "title" for the checkboxes in useDataTable hook. I added the ability to override the "selection" column if a custom title is desired and added an example to the docs.

machikoyasuda commented:
Approved 🏁 

Made a new ticket for part 2: https://github.com/reactioncommerce/catalyst/issues/110

machikoyasuda commented:
The highlight rows look good:

<img width="983" alt="Screen Shot 2019-09-24 at 3 17 42 PM" src="https://user-images.githubusercontent.com/3673236/65554454-7c3d2a00-dede-11e9-8333-ebfac8db55c8.png">

rc-publisher commented:
:tada: This PR is included in version 1.10.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@reactioncommerce/catalyst)
- [GitHub release](https://github.com/reactioncommerce/catalyst/releases/tag/v1.10.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
